TOON: Add subtitles support for cutscenes

BLooperZ opened the pull request:
This allows subtitles during cutscenes in Toonstruck.
Attaching subtitles file for first part of intro cutscene - should be extracted to `MISC` directory
[209_1M.zip](https://github.com/scummvm/scummvm/files/3721733/209_1M.zip)

Subtitles rendered using the same text draw method from the game so they look consistent.
text color changes with the palette of the cutscene.

The original game does not have subtitles in cutscenes (new feature),
I also could not find a transcript of the cutscenes' texts.
so the hard work of transcribing and timing the subtitles is still ahead.

Subtitles files are stored  in custom simple format (`tss` for Toonstruck Subtitles) consists of:
index - start frame, end frame, text offset
texts - null terminated strings

subtitles files should have the same name as matching cutscene files (minus the extension).
for example: 209_1M.SMK (cutscene) -> 209_1M.tss (subtitles).
(When all subtitles are done, I think of putting them all under a single `SUBTITLES.PAK` file)

(I have a simple python script to convert `sbv` to `tss`... I would like to add it to `devtools`.
need explanation about those `module.mk` files, though)

sev- commented on engines/toon/toon.cpp:
```diff
@@ -3286,7 +3286,18 @@ void ToonEngine::drawConversationLine() {
+	}
+}
+
+void ToonEngine::drawCustomText(int16 x, int16 y, char* line, Graphics::Surface* frame, char color) {
```
code formatting

sev- commented on engines/toon/toon.h:
```diff
@@ -212,6 +213,8 @@ class ToonEngine : public Engine {
 	void waitForScriptStep();
 	void doMagnifierEffect();
 
+	void drawCustomText(int16 x, int16 y, char* line, Graphics::Surface* frame, char color);
```
code formatting

BLooperZ commented:
Thank you for your feedback.
I have now fixed the problems.
(the obscure part is rewritten...
it's purpose is finding available byte not used in video frame to change its color for use as subtitles text color)

bluegr commented:
Nice work!

Some fixes on the subtitles:
    tommorow -> tomorrow
    I'm so DEAD! -> I'm SO dead!
    REVITILIZING -> REVITALIZING
    Here so I see it -> Here's how I see it
    A whole truckload of them is got to be -> A whole truckload of the things has got to be

And some more text afterwards (up to 5:10, around the middle of the intro):
   -  I want the new characters on my desk first thing tomorrow morning... and this time, DON'T be late!
    - Tomorrow morning, first thing, yes.
    - OK.... rabbits! More sweet, friendly, rabbits! Rabbits think nice things, rabbits say "I care". Ship!
    - You only worked with a very young, very naive Drew Blank... a kid that was going to take the world by storm... Flux. Hey there, lil' buddy. The Flux Wildly show was really going to be something... if only
    - Must... focus... on... bunnies.... Must... focus... on... bunnies! *yawn*
    - Hey cuties! Let's get hap... hap... happy! It's fluffy fluffy bun bun time!
    - Oh Drew! What are you doing at 4 o'clock in the morning anyway?
    - I've been caught with bunnies! This must be a dream! I must have fallen asleep! Oh God, sleep! I got to... Wake up Drew! Wake up! Damn it!

BLooperZ commented:
Thank you.
added support for loading `SUBTITLES.PAK` and updated subtitles file (extract into `MISC`)
[SUBTITLES.zip](https://github.com/scummvm/scummvm/files/3730968/SUBTITLES.zip)

sev- commented on engines/toon/movie.cpp:
```diff
@@ -115,9 +119,30 @@ void Movie::playVideo(bool isFirstIntroVideo) {
+					int32 currentFrame = _decoder->getCurFrame();
+
+					// find unused color key to replace with subtitles color
+					int len = frame->w * frame->h;
```
This should be uint

sev- commented on engines/toon/movie.cpp:
```diff
@@ -115,9 +119,30 @@ void Movie::playVideo(bool isFirstIntroVideo) {
+
+					// find unused color key to replace with subtitles color
+					int len = frame->w * frame->h;
+					const byte* pixels = (const byte *)frame->getPixels();
```
`const byte *pixels`

sev- commented on engines/toon/movie.cpp:
```diff
@@ -115,9 +119,30 @@ void Movie::playVideo(bool isFirstIntroVideo) {
+					const byte* pixels = (const byte *)frame->getPixels();
+					byte counts[256];
+					memset(counts, 0, sizeof(counts));
+					for (int i = 0; i < len; i++) {
```
This is also uint

sev- commented on engines/toon/movie.cpp:
```diff
@@ -103,6 +106,7 @@ void Movie::playVideo(bool isFirstIntroVideo) {
 	while (!_vm->shouldQuit() && !_decoder->endOfVideo()) {
 		if (_decoder->needsUpdate()) {
 			const Graphics::Surface *frame = _decoder->decodeNextFrame();
+			byte unused = 0;
```
maybe use of `bool` would be better?

BLooperZ replied:
this variable should later get assigned with unused value in range [1, 256] afterwards (color key) - 0 means it failed to find unused value (and already can't be unused, as it used for text border)

sev- commented on engines/toon/movie.cpp:
```diff
@@ -115,9 +119,30 @@ void Movie::playVideo(bool isFirstIntroVideo) {
+					// find unused color key to replace with subtitles color
+					int len = frame->w * frame->h;
+					const byte* pixels = (const byte *)frame->getPixels();
+					byte counts[256];
```
And `bool` here too perhaps? Because you're not counting.

sev- commented on engines/toon/movie.cpp:
```diff
@@ -115,9 +119,30 @@ void Movie::playVideo(bool isFirstIntroVideo) {
 					// WORKAROUND: There is an encoding glitch in the first intro video. This hides this using the adjacent pixels.
 					if (isFirstIntroVideo) {
-						int32 currentFrame = _decoder->getCurFrame();
+						// int32 currentFrame = _decoder->getCurFrame();
```
Why this is commented out now? If you're referring to the new call above, then don't you change the semantics, as this call was only when `isFirstIntroVideo` is set?

BLooperZ replied:
(sorry, I don't think I understand the question) this line was moved (line 122) so it's now unconditional as frame information should now be used for all videos and not just this workaround here.

sev- commented on engines/toon/subtitles.cpp:
```diff
@@ -0,0 +1,94 @@
+		if (_index > _last) {
+			return;
+		}
+		_currentLine = (char*)_fileData + _tw[_index].foffset;
```
`char *`

sev- commented on engines/toon/toon.cpp:
```diff
@@ -3286,7 +3289,18 @@ void ToonEngine::drawConversationLine() {
+	}
+}
+
+void ToonEngine::drawCustomText(int16 x, int16 y, char *line, Graphics::Surface *frame, char color) {
```
`byte color` is better, as you may get color index >0x7f, and `char` is a signed type on number of platforms.

`line` could also perhaps be better as `byte`, because you potentially could get characters with diacritics.

BLooperZ replied:
`line` matches with the type of `_currentTextLine` (in `toon.h` 391) for consistency

sev- commented on engines/toon/toon.h:
```diff
@@ -33,6 +33,7 @@
 #include "toon/state.h"
 #include "toon/picture.h"
 #include "toon/anim.h"
+#include "toon/subtitles.h"
```
Why do you need to include it here? It adds unneeded dependency.

BLooperZ commented:
Thanks again. notes marked as resolved are now fixed.

sev- commented:
After looking at it once more, I have one suggestion. Please, tell me what you think.

Current format for subtitles is basically binary. I was solving basically the same problem for Broken Sword 1 engine. You may get the subtitles pack here: https://www.scummvm.org/frs/extras/Broken%20Sword%20I%20and%20II/Sword1_Cutscenes_Subtitles-1.1.zip  and the code is here: https://github.com/scummvm/scummvm/blob/master/engines/sword1/animation.cpp#L131

The format is plaintext:
<start frame> <end frame> @<color> <subtitle text>

That could let anyone to create their custom subtitles for their language.

Do you think it worth changing the format, or still too much work?

sev- commented:
BTW, this needs to be rebased, so the merge commit goes away. Also, what questions do you have regarding the ``module.mk`` file?

BLooperZ commented:
I think plaintext format is better,  I wil try to work it into this as well, thanks.

Regarding `module.mk`, I can't find any information about it
I can see it is used to create makefile, so it allows creating  `make devtools/...` commands.
but I don't underatand how to use it and where to put it.
Is it well-known tool? if so, where can I find documentation?
BTW, If the subtitles are plaintext there is no need to add build tool (no need to add `module.mk`).

Thanks.

sev- commented:
`module.mk` is used by `makefile` utility on *nix-like systems, together with `configure`. I can add the tool for you, it is pretty straightforward. Once some tool is added, it could be built by calling `make <target-name>`, where `target-name` is normally your tool name, like `make toon-subtitles`

sev- commented:
And in general, I can take care of the `module.mk`, especially if you explain how your tool should be invoked. Also, I think that since it is written in python, there is nothing to build, and the only thing you need is to provide documentation next to it or at the beginning of the file.

bluegr commented:
I see you changed the subtitles format as per @sev- 's suggestion. Cool! :)

Could you provide a sample subtitle file with the new text format?

BLooperZ commented:
[SUBTITLES.zip](https://github.com/scummvm/scummvm/files/3827715/SUBTITLES.zip)
In this file the subtitles available both as separate plaintext files (`.tss`) or as single `.pak` file.
Each is enough to activate the subtitles.
When both available in game's director, the separate files take precedence.

sev- commented:
Nice, I'm merging this.
